From 2f0a8eb8a5763c76767fa80a9680e19ce331847d Mon Sep 17 00:00:00 2001
From: Bright Cheng <bright_cheng@wiwynn.com>
Date: Mon, 26 Aug 2024 11:37:53 +0800
Subject: [PATCH 200/204] Utils: support powerState for multi-node system

The current code of dbus-sensors only get powerState values for Chassis0
and Host0 which cannot be used for the system with multiple server
nodes.

This commit add "slotId" in Sensor class to indicate the index of
server. When setupPowerMatchCallback is called, code will call
GetSubTreePaths method to get chassis and host objects to set match
rules for each object. Meanwhile, sensor can get corresponding powerState
value with its slotId.

Test Plan:
1. Add setting "SlotId" and "PowerState" to some sensors
2. Modify property "CurrentPowerState" to
xyz.openbmc_project.State.Chassis.PowerState.Off, the sensor value of
corresponding server should be nan.
3. Change property "CurrentPowerState" to
xyz.openbmc_project.State.Chassis.PowerState.Off, the sensor value of
corresponding server should be real value.

Test Result:
1.Do Chassis power Off on Slot7
-root@bmc:~# busctl get-property xyz.openbmc_project.State.Chassis7
 /xyz/openbmc_project/state/chassis7 xyz.openbmc_project.State.Chassis
 CurrentPowerState
 s "xyz.openbmc_project.State.Chassis.PowerState.Off"
-root@bmc:~#  busctl get-property xyz.openbmc_project.PSUSensor
 /xyz/openbmc_project/sensors/voltage/MEDUSA_MB7_HSC_12V_INPUT_VOLT_V
 xyz.openbmc_project.Sensor.Value Value
 d nan
-journal log
```
Sep 17 15:15:02 bmc psusensor[2041]: Received Chassis 7 chassis power
change signal
Sep 17 15:15:02 bmc psusensor[2041]: Chassis 7 turned off
Sep 17 15:15:02 bmc psusensor[2041]: Chassis 0 CurrentPowerState : Off
Sep 17 15:15:02 bmc psusensor[2041]: Chassis 1 CurrentPowerState : On
Sep 17 15:15:02 bmc psusensor[2041]: Chassis 2 CurrentPowerState : On
Sep 17 15:15:02 bmc psusensor[2041]: Chassis 3 CurrentPowerState : On
Sep 17 15:15:02 bmc psusensor[2041]: Chassis 4 CurrentPowerState : On
Sep 17 15:15:02 bmc psusensor[2041]: Chassis 5 CurrentPowerState : On
Sep 17 15:15:02 bmc psusensor[2041]: Chassis 6 CurrentPowerState : On
Sep 17 15:15:02 bmc psusensor[2041]: Chassis 7 CurrentPowerState : Off
Sep 17 15:15:02 bmc psusensor[2041]: Chassis 8 CurrentPowerState : On
Sep 17 15:15:02 bmc psusensor[2041]: PowerStateChanged: Slot 7,
newState: Off, PowerState_type: ChassisOn
Sep 17 15:15:02 bmc psusensor[2041]: Checking slot 0, sensor MEDUSA_ADC
_48V_SENSE_VOLT_V  PowerState_type: ChassisOn
Sep 17 15:15:02 bmc psusensor[2041]: Checking slot 0, sensor MEDUSA_DEL
TA0_12V_CURR_A  PowerState_type: ChassisOn
Sep 17 15:15:02 bmc psusensor[2041]: Checking slot 6, sensor MEDUSA_MB6
_HSC_12V_TEMP_C  PowerState_type: ChassisOn
Sep 17 15:15:02 bmc psusensor[2041]: Checking slot 7, sensor MEDUSA_MB7
_HSC_12V_1_CURR_A  PowerState_type: Chas sisOn
Sep 17 15:15:02 bmc psusensor[2041]: Deactivating slot 7, sensor
MEDUSA_MB7_HSC_12V_1_CURR_A, due to PowerState::chassisOn
Sep 17 15:15:02 bmc psusensor[2041]: Checking slot 7, sensor MEDUSA_MB7
_HSC_12V_INPUT_1_PWR_W  PowerState_type: ChassisOn
Sep 17 15:15:02 bmc psusensor[2041]: Deactivating slot 7, sensor
MEDUSA_MB7_HSC_12V_INPUT_1_PWR_W, due to PowerState::chassisOn
Sep 17 15:15:02 bmc psusensor[2041]: Checking slot 7, sensor MEDUSA_MB7
_HSC_12V_INPUT_VOLT_V  PowerState_type: ChassisOn
Sep 17 15:15:02 bmc psusensor[2041]: Deactivating slot 7, sensor
MEDUSA_MB7_HSC_12V_INPUT_VOLT_V, due to PowerState::chassisOn
Sep 17 15:15:02 bmc psusensor[2041]: Checking slot 7, sensor MEDUSA_MB7
_HSC_12V_OUTPUT_VOLT_V  PowerState_type: ChassisOn
Sep 17 15:15:02 bmc psusensor[2041]: Deactivating slot 7, sensor
MEDUSA_MB7_HSC_12V_OUTPUT_VOLT_V, due to PowerState::chassisOn
Sep 17 15:15:02 bmc psusensor[2041]: Checking slot 7, sensor MEDUSA_MB7
_HSC_12V_TEMP_C  PowerState_type: ChassisOn
Sep 17 15:15:02 bmc psusensor[2041]: Deactivating slot 7, sensor
MEDUSA_MB7_HSC_12V_TEMP_C, due to PowerState:: chassisOn
```
2.Do Chassis power On on Slot7
-root@bmc:~# busctl get-property xyz.openbmc_project.State.Chassis7
 /xyz/openbmc_project/state/chassis7 xyz.openbmc_project.State.Chassis
 CurrentPowerState
 s "xyz.openbmc_project.State.Chassis.PowerState.On"
-root@bmc:~# busctl get-property xyz.openbmc_project.PSUSensor
 /xyz/openbmc_project/sensors/voltage/MEDUSA_MB7_HSC_12V_INPUT_VOLT_V
 xyz.openbmc_project.Sensor.Value
 d 12.199
-journal log
```
Sep 17 15:19:08 bmc psusensor[2041]: Received Chassis 7 chassis power
change signal
Sep 17 15:19:18 bmc psusensor[2041]: Chassis 7 turned on
Sep 17 15:19:18 bmc psusensor[2041]: Chassis 0 CurrentPowerState : Off
Sep 17 15:19:18 bmc psusensor[2041]: Chassis 1 CurrentPowerState : On
Sep 17 15:19:18 bmc psusensor[2041]: Chassis 2 CurrentPowerState : On
Sep 17 15:19:18 bmc psusensor[2041]: Chassis 3 CurrentPowerState : On
Sep 17 15:19:18 bmc psusensor[2041]: Chassis 4 CurrentPowerState : On
Sep 17 15:19:18 bmc psusensor[2041]: Chassis 5 CurrentPowerState : On
Sep 17 15:19:18 bmc psusensor[2041]: Chassis 6 CurrentPowerState : On
Sep 17 15:19:18 bmc psusensor[2041]: Chassis 7 CurrentPowerState : On
Sep 17 15:19:18 bmc psusensor[2041]: Chassis 8 CurrentPowerState : On
Sep 17 15:19:18 bmc psusensor[2041]: PowerStateChanged: Slot 7,
newState: On, PowerState_type: ChassisOn
Sep 17 15:19:18 bmc psusensor[2041]: Activated existing 'MEDUSA_MB7_HSC
_12V_OUTPUT_VOLT_V' sensor
Sep 17 15:19:18 bmc psusensor[2041]: Activated existing 'MEDUSA_MB7_HSC
_12V_1_CURR_A' sensor
Sep 17 15:19:18 bmc psusensor[2041]: Activated existing 'MEDUSA_MB7_HSC
_12V_INPUT_1_PWR_W' sensor
Sep 17 15:19:18 bmc psusensor[2041]: Activated existing 'MEDUSA_MB7_HSC
_12V_INPUT_VOLT_V' sensor
Sep 17 15:19:18 bmc psusensor[2041]: Activated existing 'MEDUSA_MB7_HSC
_12V_TEMP_C' sensor
```
Change-Id: I49ede1e6276acdbee2f8a9807c738be7970beaf9
---
 src/Utils.cpp                    | 510 ++++++++++++++++++++++---------
 src/Utils.hpp                    |  41 ++-
 src/hwmon-temp/HwmonTempMain.cpp |   6 +-
 src/psu/PSUSensor.cpp            |   6 +-
 src/psu/PSUSensor.hpp            |   8 +-
 src/psu/PSUSensorMain.cpp        |   9 +-
 src/sensor.hpp                   |   7 +-
 7 files changed, 413 insertions(+), 174 deletions(-)

diff --git a/src/Utils.cpp b/src/Utils.cpp
index 52f9394..93375b1 100644
--- a/src/Utils.cpp
+++ b/src/Utils.cpp
@@ -41,6 +41,7 @@
 #include <filesystem>
 #include <fstream>
 #include <functional>
+#include <iostream>
 #include <iterator>
 #include <memory>
 #include <optional>
@@ -56,14 +57,29 @@
 #include <variant>
 #include <vector>
 
-static bool powerStatusOn = false;
-static bool biosHasPost = false;
+using powerStatePair = std::pair<std::string, bool>;
+using powerMatchPair =
+    std::pair<std::string, std::unique_ptr<sdbusplus::bus::match_t>>;
+
 static bool manufacturingMode = false;
-static bool chassisStatusOn = false;
+static std::vector<powerStatePair> powerStatusOn;
+static std::vector<powerStatePair> biosHasPost;
+static std::vector<powerStatePair> chassisStatusOn;
+static std::vector<powerStatePair> chassisReqPowerTransOff;
 
 static std::unique_ptr<sdbusplus::bus::match_t> powerMatch = nullptr;
 static std::unique_ptr<sdbusplus::bus::match_t> postMatch = nullptr;
 static std::unique_ptr<sdbusplus::bus::match_t> chassisMatch = nullptr;
+static std::unique_ptr<sdbusplus::bus::match_t> chassisReqPowerTransOffMatch = nullptr;
+
+static std::vector<powerMatchPair> powerMatchVec;
+static std::vector<powerMatchPair> postMatchVec;
+static std::vector<powerMatchPair> chassisMatchVec;
+static std::vector<powerMatchPair> chassisReqPowerTransOffMatchVec;
+
+static std::vector<std::shared_ptr<boost::asio::steady_timer>> timerVec;
+static std::vector<std::shared_ptr<boost::asio::steady_timer>>
+    timerChassisStatusVec;
 
 /**
  * return the contents of a file
@@ -304,44 +320,105 @@ bool findFiles(const std::filesystem::path& dirPath,
     return true;
 }
 
-bool isPowerOn()
+std::vector<powerStatePair>::iterator findPowerStateByPath(
+    const std::string& path, std::vector<powerStatePair>& powerStatus)
+{
+    return std::find_if(
+        powerStatus.begin(), powerStatus.end(),
+        [&path](const auto& pair) { return pair.first == std::string(path); });
+}
+
+std::vector<powerMatchPair>::iterator findPowerMatchByPath(
+    const std::string& path, std::vector<powerMatchPair>& powerMatch)
+{
+    return std::find_if(
+        powerMatch.begin(), powerMatch.end(),
+        [&path](const auto& pair) { return pair.first == std::string(path); });
+}
+
+bool isPowerOn(const size_t& slotId)
 {
-    if (!powerMatch)
+    std::string path = power::path + std::to_string(slotId);
+    auto match_it = findPowerMatchByPath(path, powerMatchVec);
+    if (match_it == powerMatchVec.end())
     {
         throw std::runtime_error("Power Match Not Created");
     }
-    return powerStatusOn;
+    auto it = findPowerStateByPath(path, powerStatusOn);
+    if (it == powerStatusOn.end())
+    {
+        return false;
+    }
+
+    return it->second;
 }
 
-bool hasBiosPost()
+bool hasBiosPost(const size_t& slotId)
 {
-    if (!postMatch)
+    std::string path = post::path + std::to_string(slotId);
+    auto match_it = findPowerMatchByPath(path, postMatchVec);
+    if (match_it == postMatchVec.end())
     {
         throw std::runtime_error("Post Match Not Created");
     }
-    return biosHasPost;
+    auto it = findPowerStateByPath(path, biosHasPost);
+    if (it == biosHasPost.end())
+    {
+        return false;
+    }
+
+    return it->second;
 }
 
-bool isChassisOn()
+bool isChassisOn(const size_t& slotId)
 {
-    if (!chassisMatch)
+    std::string path = chassis::path + std::to_string(slotId);
+    auto match_it = findPowerMatchByPath(path, chassisMatchVec);
+    if (match_it == chassisMatchVec.end())
     {
         throw std::runtime_error("Chassis On Match Not Created");
     }
-    return chassisStatusOn;
+    auto it = findPowerStateByPath(path, chassisStatusOn);
+    if (it == chassisStatusOn.end())
+    {
+        return false;
+    }
+    return it->second;
+}
+
+bool isChassisReqTransOff(const size_t& slotId)
+{
+    std::string path = chassis::path + std::to_string(slotId);
+    auto match_it = findPowerMatchByPath(path, chassisReqPowerTransOffMatchVec);
+    if (match_it == chassisReqPowerTransOffMatchVec.end())
+    {
+        throw std::runtime_error("Chassis Requested Transition to Off Match Not Created");
+    }
+    auto it = findPowerStateByPath(path, chassisReqPowerTransOff);
+    if (it == chassisReqPowerTransOff.end())
+    {
+        return false;
+    }
+    return it->second;
 }
 
-bool readingStateGood(const PowerState& powerState)
+
+bool readingStateGood(const PowerState& powerState, const size_t& slotId)
 {
-    if (powerState == PowerState::on && !isPowerOn())
+    if (powerState == PowerState::on && !isPowerOn(slotId))
+    {
+        return false;
+    }
+    if (powerState == PowerState::biosPost &&
+        (!hasBiosPost(slotId) || !isPowerOn(slotId)))
     {
         return false;
     }
-    if (powerState == PowerState::biosPost && (!hasBiosPost() || !isPowerOn()))
+    if (powerState == PowerState::chassisOn && !isChassisOn(slotId))
     {
         return false;
     }
-    if (powerState == PowerState::chassisOn && !isChassisOn())
+    if (powerState == PowerState::chassisOn && isChassisReqTransOff(slotId))
     {
         return false;
     }
@@ -351,11 +428,15 @@ bool readingStateGood(const PowerState& powerState)
 
 static void getPowerStatus(
     const std::shared_ptr<sdbusplus::asio::connection>& conn,
-    size_t retries = 2)
+    size_t slotNumber = 0, size_t retries = 2)
 {
+    std::string busname = power::busname + std::to_string(slotNumber);
+    std::string path = power::path + std::to_string(slotNumber);
+
     conn->async_method_call(
-        [conn, retries](boost::system::error_code ec,
-                        const std::variant<std::string>& state) {
+        [conn, retries, slotNumber,
+         path](boost::system::error_code ec,
+               const std::variant<std::string>& state) {
             if (ec)
             {
                 if (retries != 0U)
@@ -363,10 +444,10 @@ static void getPowerStatus(
                     auto timer = std::make_shared<boost::asio::steady_timer>(
                         conn->get_io_context());
                     timer->expires_after(std::chrono::seconds(15));
-                    timer->async_wait(
-                        [timer, conn, retries](boost::system::error_code) {
-                            getPowerStatus(conn, retries - 1);
-                        });
+                    timer->async_wait([timer, conn, retries,
+                                       slotNumber](boost::system::error_code) {
+                        getPowerStatus(conn, slotNumber, retries - 1);
+                    });
                     return;
                 }
 
@@ -376,19 +457,24 @@ static void getPowerStatus(
                            "ERROR_MESSAGE", ec.message());
                 return;
             }
-            powerStatusOn = std::get<std::string>(state).ends_with(".Running");
+            auto it = findPowerStateByPath(path, powerStatusOn);
+            it->second = std::get<std::string>(state).ends_with(".Running");
         },
-        power::busname, power::path, properties::interface, properties::get,
-        power::interface, power::property);
+        busname, path, properties::interface, properties::get, power::interface,
+        power::property);
 }
 
 static void getPostStatus(
     const std::shared_ptr<sdbusplus::asio::connection>& conn,
-    size_t retries = 2)
+    size_t slotNumber = 0, size_t retries = 2)
 {
+    std::string busname = post::busname + std::to_string(slotNumber);
+    std::string path = post::path + std::to_string(slotNumber);
+
     conn->async_method_call(
-        [conn, retries](boost::system::error_code ec,
-                        const std::variant<std::string>& state) {
+        [conn, retries, slotNumber,
+         path](boost::system::error_code ec,
+               const std::variant<std::string>& state) {
             if (ec)
             {
                 if (retries != 0U)
@@ -396,10 +482,10 @@ static void getPostStatus(
                     auto timer = std::make_shared<boost::asio::steady_timer>(
                         conn->get_io_context());
                     timer->expires_after(std::chrono::seconds(15));
-                    timer->async_wait(
-                        [timer, conn, retries](boost::system::error_code) {
-                            getPostStatus(conn, retries - 1);
-                        });
+                    timer->async_wait([timer, conn, retries,
+                                       slotNumber](boost::system::error_code) {
+                        getPostStatus(conn, slotNumber, retries - 1);
+                    });
                     return;
                 }
                 // we commonly come up before power control, we'll capture the
@@ -409,21 +495,26 @@ static void getPostStatus(
                 return;
             }
             const auto& value = std::get<std::string>(state);
-            biosHasPost = (value != "Inactive") &&
-                          (value != "xyz.openbmc_project.State.OperatingSystem."
-                                    "Status.OSStatus.Inactive");
+            auto it = findPowerStateByPath(path, biosHasPost);
+            it->second = (value != "Inactive") &&
+                         (value != "xyz.openbmc_project.State.OperatingSystem."
+                                   "Status.OSStatus.Inactive");
         },
-        post::busname, post::path, properties::interface, properties::get,
-        post::interface, post::property);
+        busname, path, properties::interface, properties::get, post::interface,
+        post::property);
 }
 
 static void getChassisStatus(
     const std::shared_ptr<sdbusplus::asio::connection>& conn,
-    size_t retries = 2)
+    size_t slotNumber = 0, size_t retries = 2)
 {
+    std::string busname = chassis::busname + std::to_string(slotNumber);
+    std::string path = chassis::path + std::to_string(slotNumber);
+
     conn->async_method_call(
-        [conn, retries](boost::system::error_code ec,
-                        const std::variant<std::string>& state) {
+        [conn, retries, slotNumber,
+         path](boost::system::error_code ec,
+               const std::variant<std::string>& state) {
             if (ec)
             {
                 if (retries != 0U)
@@ -431,10 +522,10 @@ static void getChassisStatus(
                     auto timer = std::make_shared<boost::asio::steady_timer>(
                         conn->get_io_context());
                     timer->expires_after(std::chrono::seconds(15));
-                    timer->async_wait(
-                        [timer, conn, retries](boost::system::error_code) {
-                            getChassisStatus(conn, retries - 1);
-                        });
+                    timer->async_wait([timer, conn, retries,
+                                       slotNumber](boost::system::error_code) {
+                        getChassisStatus(conn, slotNumber, retries - 1);
+                    });
                     return;
                 }
 
@@ -445,10 +536,10 @@ static void getChassisStatus(
                     "ERROR_MESSAGE", ec.message());
                 return;
             }
-            chassisStatusOn =
-                std::get<std::string>(state).ends_with(chassis::sOn);
+            auto it = findPowerStateByPath(path, chassisStatusOn);
+            it->second = std::get<std::string>(state).ends_with(chassis::sOn);
         },
-        chassis::busname, chassis::path, properties::interface, properties::get,
+        busname, path, properties::interface, properties::get,
         chassis::interface, chassis::property);
 }
 
@@ -456,8 +547,6 @@ void setupPowerMatchCallback(
     const std::shared_ptr<sdbusplus::asio::connection>& conn,
     std::function<void(PowerState type, bool state)>&& hostStatusCallback)
 {
-    static boost::asio::steady_timer timer(conn->get_io_context());
-    static boost::asio::steady_timer timerChassisOn(conn->get_io_context());
     // create a match for powergood changes, first time do a method call to
     // cache the correct value
     if (powerMatch)
@@ -465,115 +554,234 @@ void setupPowerMatchCallback(
         return;
     }
 
-    powerMatch = std::make_unique<sdbusplus::bus::match_t>(
-        static_cast<sdbusplus::bus_t&>(*conn),
-        "type='signal',interface='" + std::string(properties::interface) +
-            "',path='" + std::string(power::path) + "',arg0='" +
-            std::string(power::interface) + "'",
-        [hostStatusCallback](sdbusplus::message_t& message) {
-            std::string objectName;
-            boost::container::flat_map<std::string, std::variant<std::string>>
-                values;
-            message.read(objectName, values);
-            auto findState = values.find(power::property);
-            if (findState != values.end())
-            {
-                bool on = std::get<std::string>(findState->second)
-                              .ends_with(".Running");
-                if (!on)
+    /* Get host paths */
+    static const int depth = 1;
+    GetSubTreePathsType hostSubTreePaths;
+
+    try
+    {
+        auto method =
+            conn->new_method_call(mapper::busName, mapper::path,
+                                  mapper::interface, mapper::subtreepaths);
+        method.append("/xyz/openbmc_project/state", depth,
+                      GetSubTreePathsType({power::interface}));
+        auto reply = conn->call(method);
+        reply.read(hostSubTreePaths);
+    }
+    catch (sdbusplus::exception_t& e)
+    {
+        std::cerr << "Error getting host subtree paths: " << e.what() << "\n";
+        return;
+    }
+
+    for (const auto& path : hostSubTreePaths)
+    {
+        size_t slotNumber =
+            std::stoi(path.substr(path.find_last_of("/host") + 1));
+        powerStatusOn.emplace_back(std::make_pair(std::string(path), false));
+        biosHasPost.emplace_back(std::make_pair(std::string(path), false));
+        auto timer =
+            timerVec.emplace_back(std::make_shared<boost::asio::steady_timer>(
+                conn->get_io_context()));
+
+        powerMatch = std::make_unique<sdbusplus::bus::match_t>(
+            static_cast<sdbusplus::bus_t&>(*conn),
+            "type='signal',interface='" + std::string(properties::interface) +
+                "',path='" + std::string(path) + "',arg0='" +
+                std::string(power::interface) + "'",
+            [hostStatusCallback, path, timer](sdbusplus::message_t& message) {
+                std::string objectName;
+                boost::container::flat_map<std::string,
+                                           std::variant<std::string>>
+                    values;
+                message.read(objectName, values);
+                auto findState = values.find(power::property);
+                if (findState != values.end())
                 {
-                    timer.cancel();
-                    powerStatusOn = false;
-                    hostStatusCallback(PowerState::on, powerStatusOn);
-                    return;
+                    bool on = std::get<std::string>(findState->second)
+                                  .ends_with(".Running");
+                    auto it = findPowerStateByPath(path, powerStatusOn);
+                    if (!on)
+                    {
+                        timer->cancel();
+                        it->second = false;
+                        hostStatusCallback(PowerState::on, it->second);
+                        return;
+                    }
+                    // on comes too quickly
+                    timer->expires_after(std::chrono::seconds(10));
+                    timer->async_wait(
+                        [hostStatusCallback, it](boost::system::error_code ec) {
+                            if (ec == boost::asio::error::operation_aborted)
+                            {
+                                return;
+                            }
+                            if (ec)
+                            {
+                                lg2::error("Timer error: '{ERROR_MESSAGE}'",
+                                           "ERROR_MESSAGE", ec.message());
+                                return;
+                            }
+                            it->second = true;
+                            hostStatusCallback(PowerState::on, it->second);
+                        });
                 }
-                // on comes too quickly
-                timer.expires_after(std::chrono::seconds(10));
-                timer.async_wait(
-                    [hostStatusCallback](boost::system::error_code ec) {
-                        if (ec == boost::asio::error::operation_aborted)
-                        {
-                            return;
-                        }
-                        if (ec)
-                        {
-                            lg2::error("Timer error: '{ERROR_MESSAGE}'",
-                                       "ERROR_MESSAGE", ec.message());
-                            return;
-                        }
-                        powerStatusOn = true;
-                        hostStatusCallback(PowerState::on, powerStatusOn);
-                    });
-            }
-        });
-
-    postMatch = std::make_unique<sdbusplus::bus::match_t>(
-        static_cast<sdbusplus::bus_t&>(*conn),
-        "type='signal',interface='" + std::string(properties::interface) +
-            "',path='" + std::string(post::path) + "',arg0='" +
-            std::string(post::interface) + "'",
-        [hostStatusCallback](sdbusplus::message_t& message) {
-            std::string objectName;
-            boost::container::flat_map<std::string, std::variant<std::string>>
-                values;
-            message.read(objectName, values);
-            auto findState = values.find(post::property);
-            if (findState != values.end())
-            {
-                auto& value = std::get<std::string>(findState->second);
-                biosHasPost =
-                    (value != "Inactive") &&
-                    (value != "xyz.openbmc_project.State.OperatingSystem."
-                              "Status.OSStatus.Inactive");
-                hostStatusCallback(PowerState::biosPost, biosHasPost);
-            }
-        });
-
-    chassisMatch = std::make_unique<sdbusplus::bus::match_t>(
-        static_cast<sdbusplus::bus_t&>(*conn),
-        "type='signal',interface='" + std::string(properties::interface) +
-            "',path='" + std::string(chassis::path) + "',arg0='" +
-            std::string(chassis::interface) + "'",
-        [hostStatusCallback = std::move(hostStatusCallback)](
-            sdbusplus::message_t& message) {
-            std::string objectName;
-            boost::container::flat_map<std::string, std::variant<std::string>>
-                values;
-            message.read(objectName, values);
-            auto findState = values.find(chassis::property);
-            if (findState != values.end())
-            {
-                bool on = std::get<std::string>(findState->second)
-                              .ends_with(chassis::sOn);
-                if (!on)
+            });
+
+        powerMatchVec.emplace_back(
+            std::make_pair(std::string(path), std::move(powerMatch)));
+
+        postMatch = std::make_unique<sdbusplus::bus::match_t>(
+            static_cast<sdbusplus::bus_t&>(*conn),
+            "type='signal',interface='" + std::string(properties::interface) +
+                "',path='" + std::string(path) + "',arg0='" +
+                std::string(post::interface) + "'",
+            [hostStatusCallback, path](sdbusplus::message_t& message) {
+                std::string objectName;
+                boost::container::flat_map<std::string,
+                                           std::variant<std::string>>
+                    values;
+                message.read(objectName, values);
+                auto findState = values.find(post::property);
+                if (findState != values.end())
                 {
-                    timerChassisOn.cancel();
-                    chassisStatusOn = false;
-                    hostStatusCallback(PowerState::chassisOn, chassisStatusOn);
-                    return;
+                    auto& value = std::get<std::string>(findState->second);
+                    auto it = findPowerStateByPath(path, biosHasPost);
+                    it->second =
+                        (value != "Inactive") &&
+                        (value != "xyz.openbmc_project.State.OperatingSystem."
+                                  "Status.OSStatus.Inactive");
+                    hostStatusCallback(PowerState::biosPost, it->second);
                 }
-                // on comes too quickly
-                timerChassisOn.expires_after(std::chrono::seconds(10));
-                timerChassisOn.async_wait([hostStatusCallback](
-                                              boost::system::error_code ec) {
-                    if (ec == boost::asio::error::operation_aborted)
+            });
+        postMatchVec.emplace_back(
+            std::make_pair(std::string(path), std::move(postMatch)));
+        getPowerStatus(conn, slotNumber);
+        getPostStatus(conn, slotNumber);
+    }
+
+    /* Get chassis paths */
+    GetSubTreePathsType chassisSubTreePaths;
+
+    try
+    {
+        auto method =
+            conn->new_method_call(mapper::busName, mapper::path,
+                                  mapper::interface, mapper::subtreepaths);
+        method.append("/xyz/openbmc_project/state", depth,
+                      GetSubTreePathsType({chassis::interface}));
+        auto reply = conn->call(method);
+        reply.read(chassisSubTreePaths);
+    }
+    catch (sdbusplus::exception_t& e)
+    {
+        std::cerr << "Error getting chassis subtree paths: " << e.what()
+                  << "\n";
+        return;
+    }
+
+    for (const auto& path : chassisSubTreePaths)
+    {
+        chassisStatusOn.emplace_back(std::make_pair(std::string(path), false));
+        chassisReqPowerTransOff.emplace_back(std::make_pair(std::string(path), false));
+
+        size_t slotNumber =
+            std::stoi(path.substr(path.find_last_of("/chassis") + 1));
+        auto timerChassisOn = timerChassisStatusVec.emplace_back(
+            std::make_shared<boost::asio::steady_timer>(
+                conn->get_io_context()));
+
+        chassisMatch = std::make_unique<sdbusplus::bus::match_t>(
+            static_cast<sdbusplus::bus_t&>(*conn),
+            "type='signal',interface='" + std::string(properties::interface) +
+                "',path='" + std::string(path) + "',arg0='" +
+                std::string(chassis::interface) + "'",
+            [hostStatusCallback, slotNumber, timerChassisOn,
+             path](sdbusplus::message_t& message) {
+                std::string objectName;
+                boost::container::flat_map<std::string,
+                                           std::variant<std::string>>
+                    values;
+                message.read(objectName, values);
+                auto findState = values.find(chassis::property);
+                if (findState != values.end())
+                {
+                    bool on = std::get<std::string>(findState->second)
+                                  .ends_with(chassis::sOn);
+                    auto it = findPowerStateByPath(path, chassisStatusOn);
+                    auto itReq = findPowerStateByPath(path, chassisReqPowerTransOff);
+                    if (!on)
                     {
+                        timerChassisOn->cancel();
+                        it->second = false;
+                        hostStatusCallback(PowerState::chassisOn, it->second);
+                        if (itReq != chassisReqPowerTransOff.end())
+                        {
+                            itReq->second = true;
+                        }
                         return;
                     }
-                    if (ec)
-                    {
-                        lg2::error("Timer error: '{ERROR_MESSAGE}'",
-                                   "ERROR_MESSAGE", ec.message());
-                        return;
+                    // on comes too quickly
+                    timerChassisOn->expires_after(std::chrono::seconds(10));
+                    timerChassisOn->async_wait(
+                        [hostStatusCallback, slotNumber,
+                         it](boost::system::error_code ec) {
+                            if (ec == boost::asio::error::operation_aborted)
+                            {
+                                return;
+                            }
+                            if (ec)
+                            {
+                                std::cerr
+                                    << "Timer error " << ec.message() << "\n";
+                                return;
+                            }
+                            it->second = true;
+                            hostStatusCallback(PowerState::chassisOn,
+                                               it->second);
+                            auto itReq = findPowerStateByPath(it->first, chassisReqPowerTransOff);
+                            if (itReq != chassisReqPowerTransOff.end())
+                            {
+                                itReq->second = false;
+                            }
+                        });
+                }
+            });
+        chassisMatchVec.emplace_back(
+            std::make_pair(std::string(path), std::move(chassisMatch)));
+        getChassisStatus(conn, slotNumber);
+
+        chassisReqPowerTransOffMatch = std::make_unique<sdbusplus::bus::match_t>(
+            static_cast<sdbusplus::bus_t&>(*conn),
+            "type='signal',interface='" + std::string(properties::interface) +
+                "',path='" + std::string(path) + "',arg0='" +
+                std::string(chassis::interface) + "'",
+            [hostStatusCallback, slotNumber, timerChassisOn,
+             path](sdbusplus::message_t& message) {
+                std::string objectName;
+                boost::container::flat_map<std::string,
+                                           std::variant<std::string>>
+                    values;
+                message.read(objectName, values);
+
+                auto findRequested = values.find(chassis::propRequestedPowerTransition);
+                if (findRequested != values.end())
+                {
+                    bool rOff = std::get<std::string>(findRequested->second)
+                                  .ends_with(chassis::rOff);
+                    auto it = findPowerStateByPath(path, chassisReqPowerTransOff);
+                    if (rOff){
+                        it->second = true;
+                    } else {
+                        it->second = false;
                     }
-                    chassisStatusOn = true;
-                    hostStatusCallback(PowerState::chassisOn, chassisStatusOn);
-                });
-            }
-        });
-    getPowerStatus(conn);
-    getPostStatus(conn);
-    getChassisStatus(conn);
+                }
+            });
+        chassisReqPowerTransOffMatchVec.emplace_back(
+            std::make_pair(std::string(path), std::move(chassisReqPowerTransOffMatch)));
+
+
+    }
 }
 
 void setupPowerMatch(const std::shared_ptr<sdbusplus::asio::connection>& conn)
diff --git a/src/Utils.hpp b/src/Utils.hpp
index 9f799f5..221645b 100644
--- a/src/Utils.hpp
+++ b/src/Utils.hpp
@@ -56,6 +56,7 @@ using ManagedObjectType =
 using GetSubTreeType = std::vector<
     std::pair<std::string,
               std::vector<std::pair<std::string, std::vector<std::string>>>>>;
+using GetSubTreePathsType = std::vector<std::string>;
 using Association = std::tuple<std::string, std::string, std::string>;
 
 inline std::string escapeName(const std::string& sensorName)
@@ -80,9 +81,10 @@ bool findFiles(const std::filesystem::path& dirPath,
                std::string_view matchString,
                std::vector<std::filesystem::path>& foundPaths,
                int symlinkDepth = 1);
-bool isPowerOn();
-bool hasBiosPost();
-bool isChassisOn();
+bool isPowerOn(const size_t& slotId = 0);
+bool hasBiosPost(const size_t& slotId = 0);
+bool isChassisOn(const size_t& slotId = 0);
+bool isChassisReqTransOff(const size_t& slotId = 0);
 void setupPowerMatchCallback(
     const std::shared_ptr<sdbusplus::asio::connection>& conn,
     std::function<void(PowerState type, bool state)>&& callback);
@@ -100,7 +102,7 @@ void createAssociation(
 void findLimits(std::pair<double, double>& limits,
                 const SensorBaseConfiguration* data);
 
-bool readingStateGood(const PowerState& powerState);
+bool readingStateGood(const PowerState& powerState, const size_t& slotId = 0);
 
 constexpr const char* configInterfacePrefix =
     "xyz.openbmc_project.Configuration.";
@@ -116,6 +118,7 @@ constexpr const char* busName = "xyz.openbmc_project.ObjectMapper";
 constexpr const char* path = "/xyz/openbmc_project/object_mapper";
 constexpr const char* interface = "xyz.openbmc_project.ObjectMapper";
 constexpr const char* subtree = "GetSubTree";
+constexpr const char* subtreepaths = "GetSubTreePaths";
 } // namespace mapper
 
 namespace properties
@@ -127,27 +130,29 @@ constexpr const char* set = "Set";
 
 namespace power
 {
-const static constexpr char* busname = "xyz.openbmc_project.State.Host0";
+const static constexpr char* busname = "xyz.openbmc_project.State.Host";
 const static constexpr char* interface = "xyz.openbmc_project.State.Host";
-const static constexpr char* path = "/xyz/openbmc_project/state/host0";
+const static constexpr char* path = "/xyz/openbmc_project/state/host";
 const static constexpr char* property = "CurrentHostState";
 } // namespace power
 
 namespace chassis
 {
-const static constexpr char* busname = "xyz.openbmc_project.State.Chassis0";
+const static constexpr char* busname = "xyz.openbmc_project.State.Chassis";
 const static constexpr char* interface = "xyz.openbmc_project.State.Chassis";
-const static constexpr char* path = "/xyz/openbmc_project/state/chassis0";
+const static constexpr char* path = "/xyz/openbmc_project/state/chassis";
 const static constexpr char* property = "CurrentPowerState";
 const static constexpr char* sOn = ".On";
+const static constexpr char* propRequestedPowerTransition = "RequestedPowerTransition";
+const static constexpr char* rOff = "Transition.Off";
 } // namespace chassis
 
 namespace post
 {
-const static constexpr char* busname = "xyz.openbmc_project.State.Host0";
+const static constexpr char* busname = "xyz.openbmc_project.State.Host";
 const static constexpr char* interface =
     "xyz.openbmc_project.State.OperatingSystem.Status";
-const static constexpr char* path = "/xyz/openbmc_project/state/host0";
+const static constexpr char* path = "/xyz/openbmc_project/state/host";
 const static constexpr char* property = "OperatingSystemState";
 } // namespace post
 
@@ -217,6 +222,19 @@ inline PowerState getPowerState(const SensorBaseConfigMap& cfg)
     return state;
 }
 
+inline size_t getSlotId(const SensorBaseConfigMap& cfg)
+{
+    size_t slotId = 0; // If there's no slotId defined, default to chassis0
+    auto findSlotId = cfg.find("SlotId");
+    if (findSlotId != cfg.end())
+    {
+        std::string readSlot =
+            std::visit(VariantToStringVisitor(), findSlotId->second);
+        slotId = std::stod(readSlot);
+    }
+    return slotId;
+}
+
 inline float getPollRate(const SensorBaseConfigMap& cfg, float dflt)
 {
     float pollRate = dflt;
@@ -259,7 +277,8 @@ struct GetSensorConfiguration :
     GetSensorConfiguration(
         std::shared_ptr<sdbusplus::asio::connection> connection,
         std::function<void(ManagedObjectType& resp)>&& callbackFunc) :
-        dbusConnection(std::move(connection)), callback(std::move(callbackFunc))
+        dbusConnection(std::move(connection)),
+        callback(std::move(callbackFunc))
     {}
 
     void getPath(const std::string& path, const std::string& interface,
diff --git a/src/hwmon-temp/HwmonTempMain.cpp b/src/hwmon-temp/HwmonTempMain.cpp
index cdd7cdc..a92c940 100644
--- a/src/hwmon-temp/HwmonTempMain.cpp
+++ b/src/hwmon-temp/HwmonTempMain.cpp
@@ -597,7 +597,11 @@ static void powerStateChanged(
         {
             if (sensor != nullptr && sensor->readState == type)
             {
-                sensor->deactivate();
+                if ((type == PowerState::chassisOn) &&
+                    (!isChassisOn(sensor->slotId)))
+                {
+                    sensor->deactivate();
+                }
             }
         }
     }
diff --git a/src/psu/PSUSensor.cpp b/src/psu/PSUSensor.cpp
index 7f17bf2..c5750e6 100644
--- a/src/psu/PSUSensor.cpp
+++ b/src/psu/PSUSensor.cpp
@@ -53,14 +53,14 @@ PSUSensor::PSUSensor(
     const std::string& sensorUnits, double factor, double max, double min,
     double offset, const std::string& label, size_t tSize, double pollRate,
     const std::shared_ptr<I2CDevice>& i2cDevice,
-    std::optional<BridgeGpio>&& bridgeGpio) :
+    std::optional<BridgeGpio>&& bridgeGpio, const size_t& slotId) :
     Sensor(escapeName(sensorName), std::move(thresholdsIn), sensorConfiguration,
-           objectType, false, false, max, min, conn, powerState),
+           objectType, false, false, max, min, conn, powerState, slotId),
     i2cDevice(i2cDevice), bridgeGpio(std::move(bridgeGpio)),
     objServer(objectServer),
     inputDev(io, path, boost::asio::random_access_file::read_only),
     waitTimer(io), path(path), sensorFactor(factor), sensorOffset(offset),
-    thresholdTimer(io)
+    thresholdTimer(io), slotId(slotId)
 {
     buffer = std::make_shared<std::array<char, 128>>();
     std::string unitPath = sensor_paths::getPathForUnits(sensorUnits);
diff --git a/src/psu/PSUSensor.hpp b/src/psu/PSUSensor.hpp
index e679dc2..e66f68c 100644
--- a/src/psu/PSUSensor.hpp
+++ b/src/psu/PSUSensor.hpp
@@ -97,7 +97,7 @@ class PSUSensor : public Sensor, public std::enable_shared_from_this<PSUSensor>
               double factor, double max, double min, double offset,
               const std::string& label, size_t tSize, double pollRate,
               const std::shared_ptr<I2CDevice>& i2cDevice,
-              std::optional<BridgeGpio>&& bridgeGpio);
+              std::optional<BridgeGpio>&& bridgeGpio, const size_t& slotId);
     ~PSUSensor() override;
     void setupRead();
     void activate(const std::string& newPath,
@@ -135,6 +135,7 @@ class PSUSensor : public Sensor, public std::enable_shared_from_this<PSUSensor>
     static constexpr double defaultSensorPoll = 1.0;
     static constexpr unsigned int defaultSensorPollMs =
         static_cast<unsigned int>(defaultSensorPoll * 1000);
+    size_t slotId;
 };
 
 class PSUProperty
@@ -142,8 +143,9 @@ class PSUProperty
   public:
     PSUProperty(std::string name, double max, double min, double factor,
                 double offset) :
-        labelTypeName(std::move(name)), maxReading(max), minReading(min),
-        sensorScaleFactor(factor), sensorOffset(offset)
+        labelTypeName(std::move(name)),
+        maxReading(max), minReading(min), sensorScaleFactor(factor),
+        sensorOffset(offset)
     {}
     ~PSUProperty() = default;
 
diff --git a/src/psu/PSUSensorMain.cpp b/src/psu/PSUSensorMain.cpp
index 53e6b17..0eb6fd4 100644
--- a/src/psu/PSUSensorMain.cpp
+++ b/src/psu/PSUSensorMain.cpp
@@ -613,6 +613,7 @@ static void createSensorsCallback(
         checkGroupEvent(directory.string(), groupEventPathList);
 
         PowerState readState = getPowerState(*baseConfig);
+        size_t readSlot = getSlotId(*baseConfig);
 
         /* Check if there are more sensors in the same interface */
         int i = 1;
@@ -1064,7 +1065,7 @@ static void createSensorsCallback(
                     readState, findSensorUnit->second, factor,
                     psuProperty.maxReading, psuProperty.minReading,
                     psuProperty.sensorOffset, labelHead, thresholdConfSize,
-                    pollRate, i2cDev, std::move(bridgeGpio));
+                    pollRate, i2cDev, std::move(bridgeGpio), readSlot);
                 sensors[sensorName]->setupRead();
                 ++numCreated;
                 lg2::debug("Created '{NUM}' sensors so far", "NUM", numCreated);
@@ -1236,7 +1237,11 @@ static void powerStateChanged(
         {
             if (sensor != nullptr && sensor->readState == type)
             {
-                sensor->deactivate();
+                if ((type == PowerState::chassisOn) &&
+                    (!isChassisOn(sensor->slotId)))
+                {
+                    sensor->deactivate();
+                }
             }
         }
     }
diff --git a/src/sensor.hpp b/src/sensor.hpp
index b5c922e..2d96ccf 100644
--- a/src/sensor.hpp
+++ b/src/sensor.hpp
@@ -75,13 +75,13 @@ struct Sensor
            const std::string& configurationPath, const std::string& objectType,
            bool isSettable, bool isMutable, const double max, const double min,
            std::shared_ptr<sdbusplus::asio::connection>& conn,
-           PowerState readState = PowerState::always) :
+           PowerState readState = PowerState::always, size_t slotId = 0) :
         name(sensor_paths::escapePathForDbus(name)),
         configurationPath(configurationPath),
         configInterface(configInterfaceName(objectType)),
         isSensorSettable(isSettable), isValueMutable(isMutable), maxValue(max),
         minValue(min), thresholds(std::move(thresholdData)),
-        dbusConnection(conn), readState(readState),
+        dbusConnection(conn), readState(readState), slotId(slotId),
         instrumentation(enableInstrumentation
                             ? std::make_unique<SensorInstrumentation>()
                             : nullptr)
@@ -126,6 +126,7 @@ struct Sensor
     double hysteresisPublish = 1.0;
     std::shared_ptr<sdbusplus::asio::connection> dbusConnection;
     PowerState readState;
+    size_t slotId;
     size_t errCount{0};
     std::unique_ptr<SensorInstrumentation> instrumentation;
 
@@ -456,7 +457,7 @@ struct Sensor
 
     bool readingStateGood() const
     {
-        return ::readingStateGood(readState);
+        return ::readingStateGood(readState, slotId);
     }
 
     void markFunctional(bool isFunctional)
-- 
2.47.3

